Add more controls

TerryGeng opened the pull request:
I managed to add "pause" into the bot, now you can pause in the middle of a song and resume from where you paused. It doesn't sound useful, but this is actually a common function of any music player. I guess I can implement this by either blocking the loop reading the stdout of ffmpeg thread so ffmpeg would stop, or by count how long the song has been played and use `-ss` to fast-forward to that part. I chose the latter.

I also try to distinguish "stop" and "clear": "stop" is to stop playing the current song and move to the next song, while "clear" is to trash the entire playlist. Now it is not confusing anymore. However, this involves a major change in the behavior of this bot, that's why I open this PL:

- add "!pause" and "!p" to pause and resume,
- now "!stop" the bot won't trash the playlist anymore, it just stops playing,
- and to trash the playlist, use "!clear",
- and "!skip" command won't remove items from the playlist or jump to somewhere, it just skip the current song,
- the removal function has been separated from "!skip" to "!rm",
- and the "jump to" function has been merged into "!p".
- just to avoid confusion, I renamed "!list" to "!listfile".

I also tried to beautify the display of the "!help". Now it looks good.

UPDATE:
I refactored the message_received part, and move all command into a separate file "command.py".
Also, I managed to implement partial match for command, now "!l", "!list", "!listfile" all work.

UPDATE:
I managed to turn db.ini file into a sqlite3 database. Sqlite3 is installed with python3 by default hence no additional dependence is required. Now since the database is properly managed by sqlite3, I'm thinking about adding more functions, like storing music library into it, able to customize description for each song in the database, keep downloaded items in the database, etc. I hope it is a good start.

The help message now:
![image](https://user-images.githubusercontent.com/2306637/74599823-5a2d8980-50c3-11ea-80ad-af401a5d9449.png)

TerryGeng commented:
Added "!filematch" command, that can add all files that match a regex pattern. I think that's useful sometimes.

Mati20187 commented:
Great job! I noticed two problems, though.
1. !playlist not working (doesn't give any errors, just nothing happens after the command)
2. When trying to add URLs in the web interface, I get "405 Method Not Allowed"
The most important thing for me is to have YouTube and SoundCloud playlists working so it would be amazing if you could fix it in your fork

TerryGeng commented:
@Mati20187 Hi, thanks for reporting these problems. I have fixed them, please have a look. Let me know if those problems still persist, or new bugs appear.

Mati20187 commented:
Unfortunately, I still have problems with !playlist. 
The playlist doesn't start playing after the download is finished, the bot downloads the 2nd track and not the 1st. To start playing I have to go to the web interface and select the song. Take a look at this video, it might be more insightful: https://youtu.be/VNOk8ZskT7c

TerryGeng commented:
@Mati20187 It seems that you are confused by "No music right now" while something is apparently playing. That's a bug, I have fixed.

As for the problem related to the order of the songs in the playlist, I think something was wrong in the launch_music and I have fixed it. Hope it will work this time. Sadly, since I'm in China, I don't smooth access to youtube and soundcloud. I use proxy to debug, but it is slow recently. 

I hope you may start the bot with argument "-v", since it will provide more debug level log of the bot itself so I can better identify the cause of the problem.

Mati20187 commented:
This is what I got with the verbose option.
`Traceback (most recent call last):
  File "/usr/lib/python3.7/configparser.py", line 788, in get
    value = d[option]
  File "/usr/lib/python3.7/collections/__init__.py", line 914, in __getitem__
    return self.__missing__(key)            # support subclasses that define __missing__
  File "/usr/lib/python3.7/collections/__init__.py", line 906, in __missing__
    raise KeyError(key)
KeyError: 'ducking'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pi/botamusique/mumbleBot.py", line 652, in <module>
    var.botamusique = MumbleBot(args)
  File "/home/pi/botamusique/mumbleBot.py", line 155, in __init__
    if var.config.getboolean("bot", "ducking"):
  File "/usr/lib/python3.7/configparser.py", line 828, in getboolean
    raw=raw, vars=vars, fallback=fallback, **kwargs)
  File "/usr/lib/python3.7/configparser.py", line 808, in _get_conv
    **kwargs)
  File "/usr/lib/python3.7/configparser.py", line 802, in _get
    return conv(self.get(section, option, **kwargs))
  File "/usr/lib/python3.7/configparser.py", line 791, in get
    raise NoOptionError(option, section)
configparser.NoOptionError: No option 'ducking' in section: 'bot'`
After adding ducking = False to my configuration.ini file, which was based on an older version of configuration.default.ini, everything works perfectly.
One cosmetic change I'd recommend is to make the web interface return to the main page after adding music through "Add URL" instead of staying on the "/post" page you can see on the 
![screenshot](https://user-images.githubusercontent.com/31066931/74860253-3dea5080-5340-11ea-9f35-107fdf871594.png).
Amazing job, man! I really appreciate your effort.

Mati20187 commented:
I found one more thing. There's something wrong with radio playback
```
2020-02-19 19:33:18,355 - root - INFO - bot: received command rbplay - 71614 by Mati
2020-02-19 19:33:18,356 - root - DEBUG - cmd: Play a station by ID
2020-02-19 19:33:18,356 - root - DEBUG - cmd: Retreiving url for station ID 71614
2020-02-19 19:33:18,363 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): www.radio-browser.info:80
2020-02-19 19:33:18,498 - urllib3.connectionpool - DEBUG - http://www.radio-browser.info:80 "GET /webservice/json/stations/byid/71614 HTTP/1.1" 200 418
2020-02-19 19:33:18,501 - root - DEBUG - cmd: Added station to playlist Antyradio Greatest
2020-02-19 19:33:18,516 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): www.radio-browser.info:80
2020-02-19 19:33:18,869 - urllib3.connectionpool - DEBUG - http://www.radio-browser.info:80 "GET /webservice/v2/json/url/71614 HTTP/1.1" 200 153
2020-02-19 19:33:18,872 - root - INFO - cmd: Found url: http://antgre-01.cdn.eurozet.pl:8008/;channel.mp3
2020-02-19 19:33:18,872 - root - INFO - cmd: add to playlist: http://antgre-01.cdn.eurozet.pl:8008/;channel.mp3
2020-02-19 19:33:18,872 - root - INFO - bot: Async download next asked
2020-02-19 19:33:18,936 - root - INFO - bot: play music http://antgre-01.cdn.eurozet.pl:8008/;channel.mp3
Traceback (most recent call last):
  File "/home/pi/botamusique/mumbleBot.py", line 663, in <module>
    var.botamusique.loop()
  File "/home/pi/botamusique/mumbleBot.py", line 515, in loop
    self.launch_music()
  File "/home/pi/botamusique/mumbleBot.py", line 305, in launch_music
    self.send_msg(var.config.get('strings', 'now_playing') + util.format_current_playing())
  File "/home/pi/botamusique/util.py", line 122, in format_current_playing
    title=media.radio.get_radio_title(
NameError: name 'media' is not defined
```
The bot crashes when trying to play radio

TerryGeng commented:
@Mati20187 Thanks for testing. And I regret not being able to go through these functions one by one by myself...
I have pushed a quick fix, please check it out.

Mati20187 commented:
The radio playback problem still persists. Here's the full log from the bot being started to it crashing.
```
2020-02-20 10:56:30,423 - root - DEBUG - Starting in DEBUG loglevel
2020-02-20 10:56:30,589 - root - DEBUG - bot: command added: joinme
2020-02-20 10:56:30,590 - root - DEBUG - bot: command added: userban
2020-02-20 10:56:30,590 - root - DEBUG - bot: command added: userunban
2020-02-20 10:56:30,590 - root - DEBUG - bot: command added: urlban
2020-02-20 10:56:30,590 - root - DEBUG - bot: command added: urlunban
2020-02-20 10:56:30,590 - root - DEBUG - bot: command added: p
2020-02-20 10:56:30,590 - root - DEBUG - bot: command added: play
2020-02-20 10:56:30,590 - root - DEBUG - bot: command added: pause
2020-02-20 10:56:30,591 - root - DEBUG - bot: command added: file
2020-02-20 10:56:30,591 - root - DEBUG - bot: command added: f
2020-02-20 10:56:30,591 - root - DEBUG - bot: command added: filematch
2020-02-20 10:56:30,591 - root - DEBUG - bot: command added: fm
2020-02-20 10:56:30,591 - root - DEBUG - bot: command added: url
2020-02-20 10:56:30,591 - root - DEBUG - bot: command added: playlist
2020-02-20 10:56:30,591 - root - DEBUG - bot: command added: radio
2020-02-20 10:56:30,592 - root - DEBUG - bot: command added: rbquery
2020-02-20 10:56:30,592 - root - DEBUG - bot: command added: rbplay
2020-02-20 10:56:30,592 - root - DEBUG - bot: command added: help
2020-02-20 10:56:30,592 - root - DEBUG - bot: command added: stop
2020-02-20 10:56:30,592 - root - DEBUG - bot: command added: clear
2020-02-20 10:56:30,592 - root - DEBUG - bot: command added: kill
2020-02-20 10:56:30,593 - root - DEBUG - bot: command added: np
2020-02-20 10:56:30,593 - root - DEBUG - bot: command added: now
2020-02-20 10:56:30,593 - root - DEBUG - bot: command added: skip
2020-02-20 10:56:30,594 - root - DEBUG - bot: command added: rm
2020-02-20 10:56:30,594 - root - DEBUG - bot: command added: listfile
2020-02-20 10:56:30,594 - root - DEBUG - bot: command added: queue
2020-02-20 10:56:30,594 - root - DEBUG - bot: command added: random
2020-02-20 10:56:30,594 - root - DEBUG - bot: command added: dropdatabase
2020-02-20 10:56:30,594 - root - INFO - bot: load playlist from previous session
2020-02-20 10:56:31,063 - root - DEBUG - bot: ducking triggered
2020-02-20 10:56:36,235 - root - INFO - bot: received command rbplay - 71614 by Mati
2020-02-20 10:56:36,239 - root - DEBUG - cmd: Play a station by ID
2020-02-20 10:56:36,240 - root - DEBUG - cmd: Retreiving url for station ID 71614
2020-02-20 10:56:36,249 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): www.radio-browser.info:80
2020-02-20 10:56:36,351 - urllib3.connectionpool - DEBUG - http://www.radio-browser.info:80 "GET /webservice/json/stations/byid/71614 HTTP/1.1" 200 417
2020-02-20 10:56:36,355 - root - DEBUG - cmd: Added station to playlist Antyradio Greatest
2020-02-20 10:56:36,368 - urllib3.connectionpool - DEBUG - Starting new HTTP connection (1): www.radio-browser.info:80
2020-02-20 10:56:36,714 - urllib3.connectionpool - DEBUG - http://www.radio-browser.info:80 "GET /webservice/v2/json/url/71614 HTTP/1.1" 200 153
2020-02-20 10:56:36,718 - root - INFO - cmd: Found url: http://antgre-01.cdn.eurozet.pl:8008/;channel.mp3
2020-02-20 10:56:36,718 - root - INFO - cmd: add to playlist: http://antgre-01.cdn.eurozet.pl:8008/;channel.mp3
2020-02-20 10:56:36,719 - root - INFO - bot: Async download next asked
2020-02-20 10:56:36,806 - root - INFO - bot: play music http://antgre-01.cdn.eurozet.pl:8008/;channel.mp3
Traceback (most recent call last):
  File "/home/pi/botamusique/mumbleBot.py", line 662, in <module>
    var.botamusique.loop()
  File "/home/pi/botamusique/mumbleBot.py", line 521, in loop
    self.launch_music()
  File "/home/pi/botamusique/mumbleBot.py", line 311, in launch_music
    self.send_msg(var.config.get('strings', 'now_playing') + util.format_current_playing())
  File "/home/pi/botamusique/util.py", line 124, in format_current_playing
    current_music["url"]),
  File "/home/pi/botamusique/media/radio.py", line 57, in get_radio_title
    response = urllib.request.urlopen(request)
  File "/usr/lib/python3.7/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.7/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/usr/lib/python3.7/urllib/request.py", line 543, in _open
    '_open', req)
  File "/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.7/urllib/request.py", line 1345, in http_open
    return self.do_open(http.client.HTTPConnection, req)
  File "/usr/lib/python3.7/urllib/request.py", line 1320, in do_open
    r = h.getresponse()
  File "/usr/lib/python3.7/http/client.py", line 1336, in getresponse
    response.begin()
  File "/usr/lib/python3.7/http/client.py", line 306, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.7/http/client.py", line 288, in _read_status
    raise BadStatusLine(line)
http.client.BadStatusLine: ICY 200 OK
```

TerryGeng commented:
@Mati20187 Fixed. Please check it out. I also added a function that can save the playlist when you shut down the bot. It will reload the playlist the next time you start the bot. Just add `save_playlist = True` in your config ini file. It avoids download music from youtube every time. I believe that is useful for you.

azlux commented:
I just come back from holidays, I will check/test the change this week.

TerryGeng commented:
Just to help you understand what has changed, I think I need to summary all updates:

1. All command parts were refactored. I moved all commands into a separate command.py, and turned them into indiviual functions. They will be registered into a dict inside the bot class. Inside the bot class, the message_received function will try to partially match all incoming command and call the function of that command.

2. The function of some command has changed. I wrote them into the help message. I think the most useful new command is "!filematch".

3. Added more controls in both commands and the web interface, including "pause" implemented by `ffmpeg -ss`.

4. Turned db.ini into a sqlite3 database. Planning to implement some features required in some issues, including music library caching, etc. I tried to save current playlist into the database and recall it the next time the bot starts. It works.

5. Auto-update part. If `auto_check_update` is set true, the bot will check for updates every time it starts.

6. Thanks to @Mati20187 , I went over the entire youtube playlist part and fixed many old bugs related to it.

7. Formatted the log. I always read logs, so I decided to make it prettier. Now I think it is well formatted. I always spent some time to catch the stderr of ffmpeg.

Mati20187 commented:
Unfortunately, playlists still aren't working correctly
```
[Feb 23 11:07:03 INFO MainThread] bot: botamusique version 5, starting...
[Feb 23 11:07:03 DEBUG MainThread] Starting in DEBUG loglevel
[Feb 23 11:07:03 INFO WebThread] Starting web interface on 192.168.0.102:8181
 * Serving Flask app "interface" (lazy loading)
[Feb 23 11:07:03 DEBUG UpdateThread] update: checking for updates...
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
[Feb 23 11:07:03 INFO WebThread]  * Running on http://192.168.0.102:8181/ (Press CTRL+C to quit)
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: joinme
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: userban
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: userunban
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: urlban
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: urlunban
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: p
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: play
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: pause
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: file
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: f
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: filematch
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: fm
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: url
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: playlist
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: radio
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: rbquery
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: rbplay
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: help
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: stop
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: clear
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: kill
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: update
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: oust
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: volume
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: duck
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: duckthres
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: duckv
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: np
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: now
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: skip
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: rm
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: listfile
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: queue
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: random
[Feb 23 11:07:04 DEBUG MainThread] bot: command added: dropdatabase
[Feb 23 11:07:04 INFO MainThread] bot: load playlist from previous session
[Feb 23 11:07:04 DEBUG UpdateThread] update: no new version found.
[Feb 23 11:07:12 INFO Thread-2] bot: received command playlist - <a href="https://www.youtube.com/playlist?list=PL1qiB_iRKQOi3gq0V1Tf5CG2o02wF0rdy">https://www.youtube.com/playlist?list=PL1qiB_iRKQOi3gq0V1Tf5CG2o02wF0rdy</a> by Mati
[Feb 23 11:07:12 DEBUG Thread-2] cmd: fetching media info from playlist url https://www.youtube.com/playlist?list=PL1qiB_iRKQOi3gq0V1Tf5CG2o02wF0rdy
[youtube:playlist] PL1qiB_iRKQOi3gq0V1Tf5CG2o02wF0rdy: Downloading webpage
[download] Downloading playlist: Music
[youtube:playlist] PL1qiB_iRKQOi3gq0V1Tf5CG2o02wF0rdy: Downloading page #1
[youtube:playlist] playlist Music: Downloading 104 videos
[download] Downloading video 1 of 104
[download] Downloading video 2 of 104
[download] Downloading video 3 of 104
[download] Downloading video 4 of 104
[download] Downloading video 5 of 104
[download] Downloading video 6 of 104
[download] Downloading video 7 of 104
[download] Downloading video 8 of 104
[download] Downloading video 9 of 104
[download] Downloading video 10 of 104
[download] Downloading video 11 of 104
[download] Downloading video 12 of 104
[download] Downloading video 13 of 104
[download] Downloading video 14 of 104
[download] Downloading video 15 of 104
[download] Downloading video 16 of 104
[download] Downloading video 17 of 104
[download] Downloading video 18 of 104
[download] Downloading video 19 of 104
[download] Downloading video 20 of 104
[download] Downloading video 21 of 104
[download] Downloading video 22 of 104
[download] Downloading video 23 of 104
[download] Downloading video 24 of 104
[download] Downloading video 25 of 104
[download] Downloading video 26 of 104
[download] Downloading video 27 of 104
[download] Downloading video 28 of 104
[download] Downloading video 29 of 104
[download] Downloading video 30 of 104
[download] Downloading video 31 of 104
[download] Downloading video 32 of 104
[download] Downloading video 33 of 104
[download] Downloading video 34 of 104
[download] Downloading video 35 of 104
[download] Downloading video 36 of 104
[download] Downloading video 37 of 104
[download] Downloading video 38 of 104
[download] Downloading video 39 of 104
[download] Downloading video 40 of 104
[download] Downloading video 41 of 104
[download] Downloading video 42 of 104
[download] Downloading video 43 of 104
[download] Downloading video 44 of 104
[download] Downloading video 45 of 104
[download] Downloading video 46 of 104
[download] Downloading video 47 of 104
[download] Downloading video 48 of 104
[download] Downloading video 49 of 104
[download] Downloading video 50 of 104
[download] Downloading video 51 of 104
[download] Downloading video 52 of 104
[download] Downloading video 53 of 104
[download] Downloading video 54 of 104
[download] Downloading video 55 of 104
[download] Downloading video 56 of 104
[download] Downloading video 57 of 104
[download] Downloading video 58 of 104
[download] Downloading video 59 of 104
[download] Downloading video 60 of 104
[download] Downloading video 61 of 104
[download] Downloading video 62 of 104
[download] Downloading video 63 of 104
[download] Downloading video 64 of 104
[download] Downloading video 65 of 104
[download] Downloading video 66 of 104
[download] Downloading video 67 of 104
[download] Downloading video 68 of 104
[download] Downloading video 69 of 104
[download] Downloading video 70 of 104
[download] Downloading video 71 of 104
[download] Downloading video 72 of 104
[download] Downloading video 73 of 104
[download] Downloading video 74 of 104
[download] Downloading video 75 of 104
[download] Downloading video 76 of 104
[download] Downloading video 77 of 104
[download] Downloading video 78 of 104
[download] Downloading video 79 of 104
[download] Downloading video 80 of 104
[download] Downloading video 81 of 104
[download] Downloading video 82 of 104
[download] Downloading video 83 of 104
[download] Downloading video 84 of 104
[download] Downloading video 85 of 104
[download] Downloading video 86 of 104
[download] Downloading video 87 of 104
[download] Downloading video 88 of 104
[download] Downloading video 89 of 104
[download] Downloading video 90 of 104
[download] Downloading video 91 of 104
[download] Downloading video 92 of 104
[download] Downloading video 93 of 104
[download] Downloading video 94 of 104
[download] Downloading video 95 of 104
[download] Downloading video 96 of 104
[download] Downloading video 97 of 104
[download] Downloading video 98 of 104
[download] Downloading video 99 of 104
[download] Downloading video 100 of 104
[download] Downloading video 101 of 104
[download] Downloading video 102 of 104
[download] Downloading video 103 of 104
[download] Downloading video 104 of 104
[download] Finished downloading playlist: Music
[youtube] 4XBB8CNHs4U: Downloading webpage
[youtube] 4XBB8CNHs4U: Downloading embed webpage
[youtube] 4XBB8CNHs4U: Refetching age-gated info webpage
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Neil Cicierega - AC/VC (Music Video) (https://www.youtube.com/watch?v=DgLS1kHiVcA)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Janis Joplin Piece Of My Heart HQ (https://www.youtube.com/watch?v=3SL0oRcD7t0)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Behemoth - Lucifer (lyric video/tekst) (https://www.youtube.com/watch?v=SkcNgQwSEbQ)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Feel It Still - Portugal. The Man - Drum Cover (https://www.youtube.com/watch?v=P4bVm5FH2mA)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Behemoth-Demigod(Radio D#$&ey Version) (https://www.youtube.com/watch?v=E92zCs4sWto)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: WOODJU - Irradiant (https://www.youtube.com/watch?v=-UfITx3pFwY)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Epic Dark Battle Music • "Where Is Your God Now" by Rok Nardin (https://www.youtube.com/watch?v=bsvzP8EO65w)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Most Epic Battle Music Ever: "Deadwood" by Really Slow Motion (https://www.youtube.com/watch?v=LO3LZohKCnE)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Ганджу - V Λ C U U M [Visual Video Production(КЛИП)] (https://www.youtube.com/watch?v=bJOEAhB7eP8)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: I'll Be There (Lyric Video) - Walk Off the Earth (https://www.youtube.com/watch?v=tRwMD-hCtjk)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Somebody That I Used to Know - Walk off the Earth (Gotye - Cover) (https://www.youtube.com/watch?v=P9mybTArlsk)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Gabriela Bee - Feelin' Alright (Official Music Video) (https://www.youtube.com/watch?v=O2Zchnm4S9U)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Karma Police - Gianni and Sarah (Walk off the Earth) (https://www.youtube.com/watch?v=yE2kANx8WdA)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Payphone - Walk off the Earth (Maroon 5 Cover) (https://www.youtube.com/watch?v=CEdD2_pdG_c)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: I'll Be There - Walk Off The Earth (VR180) (https://www.youtube.com/watch?v=zCb4yRPYpiI)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Rude - Walk off the Earth (https://www.youtube.com/watch?v=I3W3mRs4ULQ)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Somebody That I Used to Know - Walk off the Earth (https://www.youtube.com/watch?v=d9NF2edxy-M)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Girls Like You - Walk off the Earth (Maroon 5 Cover) (https://www.youtube.com/watch?v=e3PcnNiWygw)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Feelin' alright - Gabriela Bee - Drum Cover (https://www.youtube.com/watch?v=aIISZkixg2M)
[Feb 23 11:07:16 INFO Thread-2] cmd: add to playlist: Me And That Man - My Church Is Black (Official Video) (https://www.youtube.com/watch?v=4XBB8CNHs4U)
[Feb 23 11:07:16 INFO MainThread] bot: play music [url] Neil Cicierega - AC/VC (Music Video) (https://www.youtube.com/watch?v=DgLS1kHiVcA) from playlist Music by Mati
[Feb 23 11:07:16 INFO MainThread] bot: current music isn't ready, downloading...
[Feb 23 11:07:16 INFO MainThread] bot: music file existed, skip downloading /tmp/64ca032e0c918b78b87e16e206ba96ef.mp3
Traceback (most recent call last):
  File "/home/pi/botamusique/mumbleBot.py", line 713, in <module>
    var.botamusique.loop()
  File "/home/pi/botamusique/mumbleBot.py", line 572, in loop
    self.launch_music()
  File "/home/pi/botamusique/mumbleBot.py", line 344, in launch_music
    self.send_msg(util.format_current_playing())
  File "/home/pi/botamusique/util.py", line 193, in format_current_playing
    'strings', 'now_playing') % (display, thumbnail_html))
TypeError: not all arguments converted during string formatting
```

TerryGeng commented:
@Mati20187 
Nothing went wrong on my side:
![image](https://user-images.githubusercontent.com/2306637/75110399-c3377300-5668-11ea-97ab-a5ea7c72e67e.png)
The playlist I used for testing is https://www.youtube.com/watch?v=BxqYUbNR-c0&list=OLAK5uy_lTyDHpKUJ3NX_RWbTz2Crj0yB4dhB0hM4

Please check these two lines in your configuration.ini or configuration.default.ini
```
[string]
now_playing = Now playing %s<br />%s
```
If now_playing option is not as shown above, please change it into this one. There should be two `%s` in it.

Mati20187 commented:
False alarm, everything's fine with the bot. I didn't notice you made changes to configuration.default.ini. After updating my configuration.ini, everything is working as expected

TerryGeng commented:
@Mati20187 @azlux I have been thinking a while about the above problem. The final thoughts I got is, putting constants into ini files is not a good practice. I packed them up in a `constants.py`, and I hope I will keep people away. I also believe this is a good step for implementing multi-lingual support.  Just add some
```
def import_constants():
    if language == 'en':
        import constants_english as constants
    elif laguage == 'fr':
        import constants_franch as constants
    elif laguage == 'zh_cn':
        import constants_chinese as constants
```

azlux commented:
> @Mati20187 @azlux I have been thinking a while about the above problem. The final thoughts I got is, putting constants into ini files is not a good practice. In [ad7e96d](https://github.com/azlux/botamusique/commit/ad7e96d91693f9d9d07250718f1f3fc0fe90c2c7) I packed them up in a `constants.py`, and I hope I will keep people away from it. I also believe this is a good step for implementing multi-lingual support. Just add some

It's not a good way to make it, it's not constant, It's shouldn't be part of code/class. People need to modify them by changing configuration.ini file.
The config is a good place for that. Maybe the issue show us we need to add control/check on customizations done by the admin.

azlux commented on constants.py:
```diff
@@ -0,0 +1,118 @@
+class commands:
```
As I've write into the comments, I'm not agree with this part.
May be we can think about some configuration.ini check.

TerryGeng replied:
Understood. I will rollback to previous commit and implement checking function into it.

azlux commented on configuration.default.ini:
```diff
@@ -47,109 +72,23 @@ listening_port = 8181
 celestiaradio = http://celestia.aiverse.org:8000/mp3_256
 jazz = http://jazz-wr04.ice.infomaniak.ch/jazz-wr04-128.mp3 "Jazz Yeah !"
 
-[strings]
```
restore this part please.

azlux commented on command.py:
```diff
@@ -0,0 +1,621 @@
+from database import Database
+
+
+def register_all_commands(bot):
```
great change here, cleaner than me !

But same comment as config file.

azlux commented:
> I also believe this is a good step for implementing multi-lingual support. Just add some
You idea is great. But maybe we can use a configuration folder where default config is translated.

TerryGeng commented:
I add something to catch the exceptions caused by users' fault and translate those exceptions to human language to help users figure out their fault.
Just to avoid people touching configuration.default.ini, I created a configuration.example.ini and wrote some instructions into it.

Imho2000 commented:
Hi guys, when trying to add URLs in the web interface, I get "405 Method Not Allowed".. it already was fixed?

azlux commented:
Check last commit. Maybe  you can give us the webserver section on the config.
If persist, i should be better to open an issue.

TerryGeng commented:
@Imho2000 This problem has already been reported by @Mati20187 on this PR. You can scroll up a little bit to see our correspondence. It has been fixed now.
